lib/core/tasks.ts

```diff
@@ -0,0 +1,138 @@
+import assert from "assert";
+import { Resonate } from "../resonate";
+import { ErrorCodes, ResonateError } from "./errors";
+import { RetryPolicy, isRetryPolicy } from "./retry";
+import { TaskMessage, TasksSource } from "./tasksSource";
+
+export type ResumeBody = {
+  promiseId: string;
+  topLevelParent: {
+    id: string;
+    name: string;
+    version: number;
+    args: any[];
+    retryPolicy: RetryPolicy;
+  };
+};
+
+export function isResumeBody(value: unknown): value is ResumeBody {
+  return (
+    typeof value === "object" &&
+    value !== null &&
+    "promiseId" in value &&
+    typeof value.promiseId === "string" &&
+    "topLevelParent" in value &&
+    typeof value.topLevelParent === "object" &&
+    value.topLevelParent !== null &&
+    "id" in value.topLevelParent &&
+    typeof value.topLevelParent.id === "string" &&
+    "name" in value.topLevelParent &&
+    typeof value.topLevelParent.name === "string" &&
+    "version" in value.topLevelParent &&
+    typeof value.topLevelParent.version === "number" &&
+    "args" in value.topLevelParent &&
+    Array.isArray(value.topLevelParent.args) &&
+    "retryPolicy" in value.topLevelParent &&
+    typeof value.topLevelParent.retryPolicy === "object" &&
+    isRetryPolicy(value.topLevelParent.retryPolicy)
+  );
+}
+
+export class TasksHandler {
+  #resonate: Resonate;
+  #source: TasksSource;
+  #callbackPromises: Map<string, { resolve: (value: any) => void; reject: (reason?: any) => void }>;
+
+  constructor(resonate: Resonate, source: TasksSource) {
+    this.#resonate = resonate;
+    this.#source = source;
+    this.#callbackPromises = new Map();
+  }
+
+  async start(): Promise<void> {
+    for await (const task of this.#source.generator) {
+      console.log("got a task");
+      // Generator only gets us bytes back, we need to encode those into an object
+      // Don't await so we can process more events concurrently
+      await this.handleTask(task);
+    }
+  }
+
+  private async handleTask(task: TaskMessage): Promise<void> {
+    const resumeBody = await this.#resonate.store.tasks.claim(task.id, task.counter);
```

Comment:
I think we need  a fast exit here in the case where we fail to claim the task. Does the store throw in that case?

Reply:
Yes the store throws in this case, I will hande this case.